### PR TITLE
Fix the login to ECR

### PR DESCRIPTION
**Pull request by jocgir**
Version 1.28 broke the automatic login to ECR. The authentication to AWS was not processed correctly resulting in error when the token expired.

### CODE REVIEW

File: docker.go [hunk 1]
```diff
@@ -507,15 +503,28 @@ func (docker *dockerConfig) refreshImage(image string) {
...
 
-func loginToECR(account string, region string) {
-	svc := ecr.New(ecr.Options{Region: region})
+func (docker *dockerConfig) tryLoginToECR(image string) error {
+	matches, _ := reutils.MultiMatch(image, reECR)
+	account, accountOk := matches["account"]
+	if !accountOk {
+		return fmt.Errorf("%v is not an ECR image", image)
+	}
+	log.Debugf("Failed to pull %v. It is an ECR image, trying again after login to AWS ECR.", image)
```
Comment (dlevesque-coveo) [inline docker.go:512]: This log line suggests we failed to do something, but we never really tried before logging it?

Comment (mouellet-coveo) [reply docker.go:512]: I think we do in fact. From line 494, we try to get UpdateCmd which I believe tries to get the image info ( so kind of pulling) and ends here if it failed.

Comment (dlevesque-coveo) [reply docker.go:512]: Yeah, I see it outside the method, I think I'd put the log line there then though. Here in this context, there's no previous action attempted.

File: docker.go [hunk 2]
```diff
@@ -507,15 +503,28 @@ func (docker *dockerConfig) refreshImage(image string) {
 	log.Println()
 }
 
-func loginToECR(account string, region string) {
-	svc := ecr.New(ecr.Options{Region: region})
+func (docker *dockerConfig) tryLoginToECR(image string) error {
```
Comment (dlevesque-coveo) [inline docker.go:506]: I think we're doing too much in this method now.. previously this was just login to ECR. Now it's renamed with the try, but it logs something that makes sense in the context of the caller, then does the try login, then also downloads the image in there. I think it was clearer when the caller was responsible for those activities, and the login method was handling login.

File: config.go [hunk 3]
```diff
@@ -162,14 +162,14 @@ func (config TGFConfig) String() string {
 
 var cachedAwsConfig *aws.Config
 
-func (tgfConfig *TGFConfig) getAwsConfig(assumeRoleDuration time.Duration) (*aws.Config, error) {
+func (tgfConfig *TGFConfig) getAwsConfig(assumeRoleDuration time.Duration) (aws.Config, error) {
```
Comment (dlevesque-coveo) [inline config.go:165]: The downside of not using a pointer here is we lose the ability to return nil alongside an error. We are now returning a possibly pretty broken config alongside an error. The caller still has to error check in either cases, but we now return some form of config anyway. I guess the source of this change was the existence of config and _config variables handling the pointer and content of it handled separately, there's probably a way to achieve both upsides here. I know Boris was taking a look at this as well.

File: docker.go (hunk 1 above)
Comment (jocgir) [reply docker.go]: I changed the error management to make it more explicit and less verbose

File: docker.go (hunk 2 above)
Comment (jocgir) [reply docker.go:521]: I moved back the get image in the calling function

File: config.go (hunk 3 above)
Comment (jocgir) [reply config.go:165]: The problem is that AWS changed the way they are dealing with that kind of objects. Previously, they were pointers and they aren't anymore. That's not a problem to return the broken structure alongside with the error IMO.

Comment (dlevesque-coveo) [reply config.go:165]: :+1:

**Review by dlevesque-coveo — APPROVED** [review]

**Review by mouellet-coveo — APPROVED** [review]

**Pull request merged**